cdap-app-templates/cdap-etl/cdap-etl-proto/src/main/java/io/cdap/cdap/etl/proto/v2/ETLStage.java

```diff
@@ -77,6 +104,117 @@ public ETLStage upgradeStage(UpgradeContext upgradeContext) {
+    // TODO: Consider passing sort order as parameter.
+    ArtifactId newPlugin = candidates.get(candidates.size() - 1);
+    String newVersion = getUpgradedVersionString(newPlugin);
+    // If getUpgradedVersionString returns null, candidate plugin is not valid for upgrade.
+    if (newVersion == null) {
+      return plugin;
+    }
+
+    ArtifactSelectorConfig newArtifactSelectorConfig =
+      new ArtifactSelectorConfig(newPlugin.getScope().name(), newPlugin.getName(),
+                                 newVersion);
+    io.cdap.cdap.etl.proto.v2.ETLPlugin upgradedEtlPlugin =
+      new io.cdap.cdap.etl.proto.v2.ETLPlugin(plugin.getName(), plugin.getType(),
+                                              plugin.getProperties(),
+                                              newArtifactSelectorConfig);
+    return upgradedEtlPlugin;
+  }
+
+  /**
+   * Returns new valid version string for plugin upgrade if any changes are required. Returns null if no change to
+   * current plugin version.
+   * Artifact selector config only stores plugin version as string, it can be either fixed version or range.
+   * Hence, if the plugin version is fixed, replace the fixed version with newer fixed version. If it is a range,
+   * move the upper bound of the range to the newest version.
+   *
+   * @param newPlugin New candidate plugin for updating plugin artifact.
+   * @return version string to be used for new plugin. Might be fixed version/version range string depending on
+   *         current use.
+   */
+  private String getUpgradedVersionString(ArtifactId newPlugin) {
+    try {
+      ArtifactVersionRange currentVersionRange =
+        io.cdap.cdap.api.artifact.ArtifactVersionRange.parse(plugin.getArtifactConfig().getVersion());
+      if (!currentVersionRange.isExactVersion()) {
+        // Current plugin version is version range.
+        if (currentVersionRange.versionIsInRange(newPlugin.getVersion())) {
+          // Do nothing and return as is. Note that plugin scope will not change.
+          // TODO: Figure out how to change plugin scope if a newer plugin is found but in different scope.
+          return null;
+        } else if (currentVersionRange.getLower().compareTo(newPlugin.getVersion()) > 0) {
+          // Current lower version is higher than newer latest version. This should not happen.
+          LOG.warn("Invalid new plugin artifact {} for upgrade for upgrading plugin {} ", newPlugin, plugin);
+          return null;
+        } else {
+          // Increase the upper bound to latest available version.
+          ArtifactVersionRange newVersionRange =
+            new ArtifactVersionRange(currentVersionRange.getLower(), currentVersionRange.isLowerInclusive(),
+                                     newPlugin.getVersion(), /*isUpperInclusive=*/true);
+          return newVersionRange.getVersionString();
+        }
+      } else if (currentVersionRange.isExactVersion()
```

Comment:
when you have a bunch of these ifs, it's more readable to put the short checks at the top and remove a level of nesting. Something like:

```
if (currentVersionRange.isExactVersion()) {
  if (currentVersionRange.getLower().compareTo(newPlugin.getVersion()) < 0) {
    return newPlugin.getVersion().getVersion();
  }
  return null;
}
.. other logic
```